### PR TITLE
[1/n][Adjoint Module] Fix MaterialGrid design-variable attribution in the adjoint gradient

### LINKED ISSUE #1984: Incorrect adjoint gradients with overlapping design variables
When the material grids of design variables are overlapped, e.g. [the eigenmode tutorial example](https://github.com/NanoComp/meep/blob/master/python/examples/adjoint_optimization/01-Introduction.ipynb), the adjoint gradients don't agree with the finite difference gradients. 

In the aforementioned example, the gradients are correct if only one block of the design region is kept in the`` geometry`` but not both:
```
geometry=[
...
mp.Block(center=design_region.center, size=design_region.size, material=design_variables), # design region
mp.Block(center=design_region.center, size=design_region.size, material=design_variables,
             e1=mp.Vector3(x=-1).rotate(mp.Vector3(z=1), np.pi/2), e2=mp.Vector3(y=1).rotate(mp.Vector3(z=1), np.pi/2))
]
```

### PR DESCRIPTION
**Pull request by smartalecH**
### Stack

Each PR is based on the one above it, so every diff here is just its own delta.

| | PR | |
|---:|---|---|
| 1 | #3277 | Fix MaterialGrid design-variable attribution (#1984) **&larr; you are here** |
| 2 | #3290 | Fix source deposition across chunk boundaries |
| 3 | #3292 | Fix adjoint source weighting on a symmetry plane (#3291) |
| 4 | #3280 | Enable proper JAX support in the adjoint module |
| 5 | #3281 | Differentiable angular-spectrum propagation |
| 6 | #3283 | Adjoint gradients w.r.t. source amplitudes |
| 7 | #3284 | Launch a mode through a stratified stack |
| 8 | #3285 | Adjoint gradients w.r.t. geometry (centre and size) |
| 9 | #3286 | Register a material grid's susceptibilities |
| 10 | #3287 | Evaluate the dispersive adjoint at the discrete lineshape |
| 11 | #3293 | Subpixel-smooth dispersive materials and their shape derivative |
| 12 | #3294 | Scalable Pade extrapolation for DFT monitors (#3217) |
| 13 | #3295 | 3D grating coupler example: fiber, grating and mirror |

Fixes #1984.

Stacked on #3274 — based on `fix/adjoint-chunk-pairing-minimal`, so the diff here is just these two commits. Merge #3274 first, then this retargets to `master` cleanly.

While working on a new 3D grating coupler TO example for meep (2 design regions) I noticed that the gradients were rather inaccurate. Claude helped me trace things down to the unique case where we have two adjacent design regions (e.g. a partial etched layer). There were different issues depending on the sizes of the design regions (e.g. if one design region was bigger than the other, then we got a segfault!)

The core issue was that the logic we had for handling multiple design grids during the recombination step wasn't properly bookkeeping the effects near the boundaries. This was because we weren't keeping track of "which grid was which." In addition, we had some hairy "copies" of grids we were trying to use to keep track of multiple grids (hence the crash). So the resulting gradient near the boundary was completely off. To fix this, we simply add a grid id (automatically populated in python) and propagated to the cpp backend via swig.

Practically speaking, this means that **the unit of differentiation is now the design variable, not the grid copy.**

Turns out this bug also affects symmetries (see #1984) and the fix also resolves it.

### CODE REVIEW

**Comment by lxvm** [comment]
I just tested this pr on a two-layer vertical grating coupler with partial etch layer calculation and it dramatically lowered the adjoint error along a directional derivative to 0.2% compared to finite differences. In #3274 the relative error was around 26% and in the latest release the relative error was around 300%. I am a bit surprised that the error dropped so much given that the discretization error for the forward calculation is on the order of 20% at the resolution I am using. Nonetheless, I am looking forward to trying out this pr in an optimization

**Comment by jin-castle** [comment]
This is really nice work, especially the symmetry cases. I ran into a similar, though distinct, issue with symmetries while validating #3274, and it was very difficult to diagnose. I really appreciate the rotated and mirrored regression coverage here.

**Comment by smartalecH** [comment]
Glad to hear it, thanks @lxvm!

> I am a bit surprised that the error dropped so much given that the discretization error for the forward calculation is on the order of 20% at the resolution I am using.

Which "resolution" is this? The simulation resolution, or the "resolution" of your directional derivative (e.g. your Δx)?

Note that meep's hybrid time-/frequency-domain adjoint should match finite difference gradients _regardless of the discretization_. That's because we are differentiating the discretized operator, not the continuous operator (e.g.  differentiate then optimize vs optimize then differentiate). Of course, the step of your finite difference matters (as does whether it's one sided vs centered). But you should get pretty good agreement whether your resolution is 10 or 1000.

We haven't always, and that's because we have bugs 🙃  but let's see if we can track those down now that we have some additional support from coding tools.

**Comment by lxvm** [comment]
Thanks @smartalecH !

The resolution I was referring to was the simulation resolution (i.e. 30 for my simulation for a coupler based on 500nm x 220 nm waveguide dimensions). Thanks for your comment because I had thought meep was using a differentiate then discretize approach, which is not the case. Glad to hear that I should expect good agreement of the adjoint gradients with finite differences (in my test I was using a finite difference step size of 10^-5).

**Comment by smartalecH** [comment]
@lxvm did you try running an optimization after rebasing on this PR?

**Comment by lxvm** [comment]
Yes, I ran an optimization with this pr and it gave improved results compared to the prs leading up to this one. I've been able to do robust optimization of a single mode coupler with length scale and partial etch constraints that attained 50% efficiency which is the best result I've had so far

**Comment by stevengj** [comment]
This looks reasonable to me.

**Comment by stevengj** [comment]
Failing test?

File: src/meepgeom.cpp [hunk 1]
```diff
@@ -2863,6 +2863,7 @@ void material_grids_addgradient_point(double *v, vector3 p, double scalegrad, ge
     do {
       material_data *mg_cur = (material_data *)tpc->objects[oic].o->material;
       if (owner_grid_id < 0 || mg_cur->grid_id == owner_grid_id) {
+        if (udatas.empty()) sz = mg_cur->grid_size;
```
Comment (smartalecH) [inline src/meepgeom.cpp:2866]: cc @stevengj claude picked this up during the rebase (another corner case). It added a test (above) which had a tolerance slightly too tight for single precision.

File: src/meepgeom.cpp [hunk 2]
```diff
@@ -2716,58 +2731,66 @@ With the addition of subpixel smoothing, however, the vJp became
...
+                          dz);
+    int x_list[2] = {x1, x2}, y_list[2] = {y1, y2}, z_list[2] = {z1, z2};
+    int lx = (x1 == x2) ? 1 : 2;
+    int ly = (y1 == y2) ? 1 : 2;
+    int lz = (z1 == z2) ? 1 : 2;
+
+    if (k == 0 && (ukind == material_data::U_MIN || ukind == material_data::U_PROD)) {
+      /* U_MIN and U_PROD carry adjustments defined in terms of the interpolated
+         value at this point. Overlapping grids of these kinds are rejected by
+         the caller, so there is exactly one copy and this is unchanged. */
```
Comment (stevengj) [inline src/meepgeom.cpp:2762]: I don't understand this comment.  Why are overlapping `U_MIN` and `U_PROD` grids "rejected by the caller"?

**Comment by smartalecH** [comment]
Closing in favour of #3297, which is the same branch and the same content, opened from this repository rather than from a fork.

The reason is purely mechanical: this series is a stack of 13 pull requests, each based on the one below it so that every diff is only its own delta. GitHub's native stacked pull requests cannot include a PR opened from a fork, and four of the thirteen — including this one — were. Rather than leave the series half-stacked, all four have been reopened with their branches hosted here so the whole chain can be linked.

Nothing about the change itself has been revised, and the description is carried over verbatim. The discussion and reviews on this PR stay readable here; please continue on #3297.
